src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala

```diff
@@ -0,0 +1,780 @@
+        case _                                 => true
+      }
+      .flatMap { case Seq(t, c) =>
+        val ctxs                 = Seq(t, c).flatten // NB: can be 1 or 2 contexts here
+        val inTruthPhaseBlock    = ctxs.exists(truthPhaseBlockDetector.overlapsAny)
+        val inCallPhaseBlock     = ctxs.exists(callPhaseBlockDetector.overlapsAny)
+        val isTruthVariant       = t.isDefined
+        val isCallVariant        = c.isDefined
+        val truthPhasedVariant   = t.filter { ctx => PhasingUtils.getPhasingSetId(ctx) > 0 }
+        val callPhasedVariant    = c.filter { ctx => PhasingUtils.getPhasingSetId(ctx) > 0 }
+        val isTruthVariantPhased = truthPhasedVariant.isDefined
+        val isCallVariantPhased  = callPhasedVariant.isDefined
+
+        // DEBUG: make sure we overlap at most one phase block
+        require(ctxs.forall(truthPhaseBlockDetector.getOverlaps(_).size <= 1))
+        require(ctxs.forall(callPhaseBlockDetector.getOverlaps(_).size <= 1))
+
+        // Fill in the metrics
+        // Basic metrics
+        if (isCallVariant) {
+          metric.num_called += 1
+          if (isCallVariantPhased) metric.num_phased += 1
+        }
+        if (isTruthVariant) {
+          metric.num_truth += 1
+          if (isTruthVariantPhased) metric.num_truth_phased += 1
+        }
+        // Less fun to compute metrics
+        if (isCallVariant && isTruthVariantPhased) {
+          metric.num_called_with_truth_phased += 1
+          if (isCallVariantPhased) metric.num_phased_with_truth_phased += 1
+        }
+        if (inCallPhaseBlock) {
+          if (isTruthVariantPhased) {
+            metric.num_truth_phased_in_called_block += 1
+            if (isCallVariantPhased) metric.num_both_phased_in_called_block += 1
+          }
+        }
+
+        // NB: only create the cigar from phased variants.
+        if (isCallVariantPhased || isTruthVariantPhased) {
+          contextsToCigar(truthPhasedVariant, callPhasedVariant)
+        }
+        else Seq.empty
+      }.toList match { // remove leading `*End`
+        case Nil            => Nil
+        case BothEnd :: xs  => xs
+        case CallEnd :: xs  => xs
+        case TruthEnd :: xs => xs
+        case xs             => xs // unreachable?
+     }
+
+    // Update the various rates...
+    metric.finalizeMetric()
+
+    Seq(BothEnd) ++ cigar ++ Seq(BothEnd)
+  }
+
+ /** Gets all the indices in the cigar where there is a short switch error.
+  *
+  * Short switches are sites where the alleles are misphased (flipped), and both upstream and downstream, either the
+  * haplotype block ends, or the next variant has the same phase.
+  */
+  def toShortSwitchErrorIndices(cigar: Cigar): Seq[Int] = {
+    // check if we can find a match or the end of a block at the given idx.  If we cannot, move to the next idx.
+    @tailrec
+    def checkNext(cigar: Seq[CigarType], idx: Int, by: Int): Boolean = {
+      if (idx < 0 || cigar.length <= idx) return true // no more
```

Comment:
This is a little belt-and-suspenders again.  You stick a `BothEnd` at each end of the Cigar, so this should never happen.